### PR TITLE
feat: Add token introspection endpoint

### PR DESCRIPTION
**Pull request by anku255**
## Summary of change

This PR - 

1. Adds the token introspection endpoint
2. Adds support for basic auth in the `/token` endpoint
3. Adds `getBodyAsJSONOrFormData` method in `BaseRequest` that is used for `/token`, `/revoke` and `/introspect` endpoints.

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [ ] Had run `npm run build-pretty`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/configUtils.ts` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.
-   [ ] If added a new entry point, then make sure that it is importable by adding it to the `exports` in `package.json`

### CODE REVIEW

File: lib/ts/framework/request.ts [hunk 1]
```diff
@@ -63,4 +63,26 @@ export abstract class BaseRequest {
         }
         return this.parsedJSONBody;
     };
+
+    getBodyAsJSONOrFormData = async (): Promise<any> => {
+        const contentType = this.getHeaderValue("content-type") || this.getHeaderValue("Content-Type");
```
Comment (porcellus) [inline lib/ts/framework/request.ts:68]: I think the casing should be normalized by the `getHeaderValue`.

File: lib/ts/framework/request.ts [hunk 2]
```diff
@@ -63,4 +63,26 @@ export abstract class BaseRequest {
...
+
+        if (contentType) {
+            if (contentType.includes("application/json")) {
```
Comment (porcellus) [inline lib/ts/framework/request.ts:71]: Shouldn't these use `startsWith`?

File: lib/ts/recipe/oauth2provider/api/token.ts [hunk 3]
```diff
@@ -26,9 +26,13 @@ export default async function tokenPOST(
         return false;
     }
 
+    const authorizationHeader =
+        options.req.getHeaderValue("authorization") || options.req.getHeaderValue("Authorization");
```
Comment (porcellus) [inline lib/ts/recipe/oauth2provider/api/token.ts:30]: We do not need both casings

**Review by porcellus — CHANGES_REQUESTED** [review]

File: lib/ts/framework/request.ts (hunk 1 above)
Comment (anku255) [reply lib/ts/framework/request.ts]: done

File: lib/ts/framework/request.ts (hunk 2 above)
Comment (anku255) [reply lib/ts/framework/request.ts]: done

File: lib/ts/recipe/oauth2provider/api/token.ts (hunk 3 above)
Comment (anku255) [reply lib/ts/recipe/oauth2provider/api/token.ts]: done

File: lib/ts/recipe/openid/api/getOpenIdDiscoveryConfiguration.ts [hunk 4]
```diff
@@ -37,6 +37,7 @@ export default async function getOpenIdDiscoveryConfiguration(
             authorization_endpoint: result.authorization_endpoint,
             token_endpoint: result.token_endpoint,
             userinfo_endpoint: result.userinfo_endpoint,
```
Comment (porcellus) [inline lib/ts/recipe/openid/api/getOpenIdDiscoveryConfiguration.ts:39]: Please add `token_introspection_endpoint`

**Review by porcellus — CHANGES_REQUESTED** [review]
There are some merge issues as well

File: lib/ts/recipe/openid/api/getOpenIdDiscoveryConfiguration.ts (hunk 4 above)
Comment (anku255) [reply lib/ts/recipe/openid/api/getOpenIdDiscoveryConfiguration.ts:39]: done

File: lib/ts/recipe/oauth2provider/recipeImplementation.ts [hunk 5]
```diff
@@ -258,6 +258,13 @@ export default function getRecipeInterface(
                 input.userContext
             );
 
+            if (res.status !== "OK") {
+                return {
+                    statusCode: res.statusCode,
+                    error: res.data.error,
+                    errorDescription: res.data.error_description,
+                };
+            }
```
Comment (porcellus) [inline lib/ts/recipe/oauth2provider/recipeImplementation.ts:267]:
```suggestion
```

File: lib/build/recipe/oauth2provider/recipeImplementation.js [hunk 6]
```diff
@@ -280,6 +280,13 @@ function getRecipeInterface(querier, _config, appInfo, getDefaultIdTokenPayload,
                     errorDescription: res.data.error_description,
                 };
             }
+            if (res.status !== "OK") {
+                return {
+                    statusCode: res.statusCode,
+                    error: res.data.error,
+                    errorDescription: res.data.error_description,
+                };
+            }
```
Comment (porcellus) [inline lib/build/recipe/oauth2provider/recipeImplementation.js:289]:
```suggestion
```

**Review by porcellus — CHANGES_REQUESTED** [review]

**Pull request merged**